ingest/clusters.py

```diff
@@ -0,0 +1,87 @@
+import copy
+import unicodedata
+from typing import Dict, Generator, List, Tuple, Union
+
+from ingest_files import IngestFiles
+
+
+class Clusters(IngestFiles):
+
+    ALLOWED_FILE_TYPES = ['text/csv',
+                          'text/plain', 'text/tab-separated-values']
+    COLLECTION_NAME = 'clusters'
+    SUBCOLLECTION_NAME = 'data'
+
+    def __init__(self, file_path, *, name: str = None, study_accession: str = "", domain_ranges=None):
+
+        IngestFiles.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
+        self.header = self.get_next_line(increase_line_count=False)
+        # Second line in cluster is metadata_type
+        self.metadata_types = self.get_next_line(increase_line_count=False)
+        self.unique_values = dict.fromkeys(self.header[1:], [])
+        self.source_file_type = 'cluster'
+        self.has_z = 'z' in self.header
+        self.top_level_doc = {
+            'name': name,
+            'study_accession': study_accession,
+            'source_file_type': 'cluster',
+            'domain_ranges': domain_ranges,
+            'points': self.amount_of_lines,
+        }
+        self.annotation_subdocs = self.create_annotation_subdocs()
+        # sample size needs to be smaller than amount of points
+        # for each cell annotations if there are any
+
+    def update_points(self):
+        self.top_level_doc['points'] = self.amount_of_lines
+
+    def transform(self, row):
+        """ Add data from cluster files into annotation subdocs in cluster data model"""
+
+        for idx, column in enumerate(row):
+            annotation = self.header[idx].casefold()
+            # first index is cell name don't need to check annot type
+            if idx != 0:
+                if self.metadata_types[idx].casefold() == 'numeric':
+                    column = round(float(column), 3)
+                elif self.metadata_types[idx].casefold() == 'group':
+                    if column not in self.unique_values[annotation]:
+                        self.unique_values[annotation].append(column)
+            # perform a shallow copy
+            annotation_value = copy.copy(
+                self.annotation_subdocs[annotation]['value'])
+            annotation_value.append(column)
+            self.annotation_subdocs[annotation]['value'] = annotation_value
+
+    def create_annotation_subdocs(self):
+        """Creates annotation_subdocs"""
+        annotation_subdocs = {}
+        for annot_name in self.header:
+            annot_name = value.casefold()
+            if annot_name == 'name':
+                annotation_subdocs[value] = self.create_metadata_subdoc(
+                    'text', 'cells')
+            elif value in ('x', 'y', 'z'):
+                annotation_subdocs[value] = self.create_metadata_subdoc(
+                    annot_name, 'coordinates')
+            else:
+                annotation_subdocs[value] = self.create_metadata_subdoc(
+                    annot_name, 'annotations')
```

Comment:
this still isn't correct - it would result in the same issue @eweitz noted earlier.  If we had an annotation header value of "Cell Type", this would get saved in the sub document as "cell type".  The else should read like this:

``
else:
    annotation_subdocs[value] = self.create_metadata_subdoc(value, 'annotations')
``

Reply:
Resolved in d64a62e.